gpbft/chain.go

```diff
@@ -1,61 +1,31 @@
 package gpbft
 
 import (
-	"encoding/binary"
+	"bytes"
+	"encoding/hex"
 	"errors"
-	"io"
-	"strconv"
 	"strings"
 )
 
-// Information about a tipset that is relevant to the F3 protocol.
-// This is a lightweight value type comprising 3 machine words.
-// Fields are exported for CBOR generation, but are opqaue and should not be accessed
-// within the protocol implementation.
-type TipSet struct {
-	// The epoch of the blocks in the tipset.
-	Epoch int64
-	// The CID of the tipset.
-	CID TipSetID
-}
-
-// Creates a new tipset.
-func NewTipSet(epoch int64, cid TipSetID) TipSet {
-	return TipSet{
-		Epoch: epoch,
-		CID:   cid,
-	}
-}
-
-// Returns a zero value tipset.
-// The zero value is not a meaningful tipset and may be used to represent bottom.
-func ZeroTipSet() TipSet {
-	return TipSet{}
-}
-
-func (t TipSet) IsZero() bool {
-	return t.Epoch == 0 && t.CID.IsZero()
-}
-
-func (t TipSet) String() string {
-	var b strings.Builder
-	b.Write(t.CID.Bytes())
-	b.WriteString("@")
-	b.WriteString(strconv.FormatInt(t.Epoch, 10))
-	return b.String()
-}
-
-func (t TipSet) MarshalForSigning(w io.Writer) {
-	_ = binary.Write(w, binary.BigEndian, t.Epoch)
-	_, _ = w.Write(t.CID.Bytes())
-}
+// Opaque type representing a tipset.
+// This is expected to be:
+// - a canonical sequence of CIDs of block headers identifying a tipset,
+// - a commitment to the resulting power table,
+// - a commitment to additional derived values.
+// However, GossipPBFT doesn't need to know anything about that structure.
+type TipSet = []byte
```

Comment:
I am curious: why type alias instead of type definition?

Reply:
cbor-gen was unable to generate correct code for [][]byte. A type definition would have been slightly better, but this isn't a yak I'm prepared to shave.